packages/core/src/hooks/useQueryParams.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import qs from 'qs';
+import { useState } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
+import { useDebounce } from 'react-use';
+
+function stringify(queryParams: any): string {
+  // Even though these setting don't look nice (e.g. escaped brackets), we should keep
+  // them this way. The current syntax handles all cases, including variable types with
+  // arrays or strings.
+  return qs.stringify(queryParams, {
+    strictNullHandling: true,
+  });
+}
+
+function parse(queryString: string): any {
+  return qs.parse(queryString, {
+    ignoreQueryPrefix: true,
+    strictNullHandling: true,
+  });
+}
+
+type SetQueryParams<T> = (params: T) => void;
+
+export function useQueryParams<T>(): [T, SetQueryParams<T>] {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const [queryParams, setQueryParams] = useState(parse(location.search));
```

Comment:
I have to admit that I thought it would work. The hook should also be able to do that. The problem is that right now the page reloads if I change a query parameter. I'm not sure if that would work at all 🤔

Reply:
I played a bit around with it, also with the existing implementation in cost insights and I think it's a general problem. One is not allowed to edit the URL manually. That will always force a reload. Do you have another idea?

Reply:
Afaik it should be possible to update query params without a reload as long as you use the html5 routing API, which we'll be using through react router and afaik `useNavigate`

Reply:
I now understand what Ben means, I only parse the location once and use it as the initial parameter for the state. Editing the URL manually without reload will never be possible (only via the history api), but that was not what he meant 😉  Right now it's more a hypothetical error, but I should fix it! 
I could use a effect for the location:
```typescript
useEffect(() => {
    setQueryParams(parse(location.search));
}, [location]);
```

Reply:
Gave this another try. Using the effect as above we get into race conditions and/or infinit recursion because I use debouncing. Not sure how to solve this right now. But for now this isn't an issue, so I would prefer to fix this later.